### PR TITLE
New infrastructure for metrics documentation

### PR DESCRIPTION
**Pull request by neunhoef**
This creates a new infrastructure for metrics documentation.
From now on, every metrics

  - MUST BE declared
  - MUST HAVE a documentation snippet

and this is enforced at compile and PR test time.

### Scope & Purpose

*(Please describe the changes in this PR for reviewers - **mandatory**)*

- [*] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)

#### Backports:

- [*] No backports required

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [*] Design document: https://github.com/arangodb/documents/blob/master/DesignDocuments/01_IDEAS/MetricsClassification.md

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [*] Added to `README_maintainers.md`

### CODE REVIEW

**Review by neunhoef — COMMENTED** [review]
LGTM

File: README_maintainers.md [hunk 1]
```diff
@@ -76,6 +76,50 @@ compiler is for C/C++. You can invoke it like this:
...
+```
+  DECLARE_COUNTER(name, helpstring);
+  DECLARE_GAUGE(name, type, helpstring);
+  DECLARE_HISTOGRAM(name, scaletype, helpstring);
+```
+
+in some `.cpp` file (please put only one on a line). Then, when the
+metric is actually requested in the source code, it gets a template
+added to the metrics feature with the `add` method and its name.
+Labels can be added with the `withLabels` method. In this way, the
```
Comment (mpoeter) [inline README_maintainers.md:94]:
```suggestion
Labels can be added with the `withLabel` method. In this way, the
```

File: arangod/Agency/Agent.cpp [hunk 2]
```diff
@@ -50,6 +50,24 @@ using namespace arangodb::application_features;
 using namespace arangodb::velocypack;
 using namespace std::chrono;
 
+
+struct AppendScale {
+  static log_scale_t<float> scale() { return {2.f, 0.f, 1000.f, 10}; }
+};
+struct AgentScale {
+  static log_scale_t<float> scale() { return {1.f, 0.f, 200.f, 10}; }
```
Comment (mpoeter) [inline arangod/Agency/Agent.cpp:58]: The old scale definition was different:
```suggestion
  static log_scale_t<float> scale() { return {std::exp(1.f), 0.f, 200.f, 10}; }
```

File: arangod/Cluster/MaintenanceFeature.cpp [hunk 3]
```diff
@@ -56,6 +56,36 @@ using namespace arangodb::application_features;
...
+DECLARE_COUNTER(arangodb_maintenance_action_duplicate_counter, "Counter of actions that have been discarded because of a duplicate");
+DECLARE_COUNTER(arangodb_maintenance_action_registered_counter, "Counter of actions that have been registered in the action registry");
+DECLARE_COUNTER(arangodb_maintenance_action_accum_runtime_msec, "Accumulated action runtime");
+DECLARE_COUNTER(arangodb_maintenance_action_accum_queue_time_msec, "Accumulated action queue time");
+DECLARE_COUNTER(arangodb_maintenance_action_failure_counter, "Failure counter for the maintenance actions");
+
+struct MaintenanceScale {
+  static log_scale_t<uint64_t> scale() { return {2, 50, 8000, 10}; }
+};
+struct MaintenanceScale2 {
```
Comment (mpoeter) [inline arangod/Cluster/MaintenanceFeature.cpp:71]:
```suggestion
struct MaintenanceActionRuntimeScale {
```

File: arangod/Cluster/MaintenanceFeature.cpp [hunk 4]
```diff
@@ -56,6 +56,36 @@ using namespace arangodb::application_features;
...
+  static log_scale_t<uint64_t> scale() { return {4, 82, 86400000, 10}; }
+};
+struct MaintenanceScale3 {
+  static log_scale_t<uint64_t> scale() { return {2, 82, 3600000, 12}; }
+};
+
+DECLARE_HISTOGRAM(arangodb_maintenance_phase1_runtime_msec, MaintenanceScale, "Maintenance Phase 1 runtime histogram [ms]");
+DECLARE_HISTOGRAM(arangodb_maintenance_phase2_runtime_msec, MaintenanceScale, "Maintenance Phase 2 runtime histogram [ms]");
+DECLARE_HISTOGRAM(arangodb_maintenance_agency_sync_runtime_msec, MaintenanceScale, "Total time spend on agency sync [ms]");
+DECLARE_HISTOGRAM(arangodb_maintenance_action_runtime_msec, MaintenanceScale2, "Time spent executing a maintenance action [ms]");
```
Comment (mpoeter) [inline arangod/Cluster/MaintenanceFeature.cpp:81]:
```suggestion
DECLARE_HISTOGRAM(arangodb_maintenance_action_runtime_msec, MaintenanceActionRuntimeScale , "Time spent executing a maintenance action [ms]");
```

File: arangod/Cluster/MaintenanceFeature.cpp [hunk 5]
```diff
@@ -56,6 +56,36 @@ using namespace arangodb::application_features;
...
+DECLARE_COUNTER(arangodb_maintenance_action_accum_queue_time_msec, "Accumulated action queue time");
+DECLARE_COUNTER(arangodb_maintenance_action_failure_counter, "Failure counter for the maintenance actions");
+
+struct MaintenanceScale {
+  static log_scale_t<uint64_t> scale() { return {2, 50, 8000, 10}; }
+};
+struct MaintenanceScale2 {
+  static log_scale_t<uint64_t> scale() { return {4, 82, 86400000, 10}; }
+};
+struct MaintenanceScale3 {
```
Comment (mpoeter) [inline arangod/Cluster/MaintenanceFeature.cpp:74]:
```suggestion
struct MaintenanceActionQueueTimeScale {
```

File: arangod/Cluster/MaintenanceFeature.cpp [hunk 6]
```diff
@@ -56,6 +56,36 @@ using namespace arangodb::application_features;
...
+  static log_scale_t<uint64_t> scale() { return {2, 82, 3600000, 12}; }
+};
+
+DECLARE_HISTOGRAM(arangodb_maintenance_phase1_runtime_msec, MaintenanceScale, "Maintenance Phase 1 runtime histogram [ms]");
+DECLARE_HISTOGRAM(arangodb_maintenance_phase2_runtime_msec, MaintenanceScale, "Maintenance Phase 2 runtime histogram [ms]");
+DECLARE_HISTOGRAM(arangodb_maintenance_agency_sync_runtime_msec, MaintenanceScale, "Total time spend on agency sync [ms]");
+DECLARE_HISTOGRAM(arangodb_maintenance_action_runtime_msec, MaintenanceScale2, "Time spent executing a maintenance action [ms]");
+DECLARE_HISTOGRAM(arangodb_maintenance_action_queue_time_msec, MaintenanceScale3, "Time spent in the queue before execution for maintenance actions [ms]");
```
Comment (mpoeter) [inline arangod/Cluster/MaintenanceFeature.cpp:82]:
```suggestion
DECLARE_HISTOGRAM(arangodb_maintenance_action_queue_time_msec, MaintenanceActionQueueTimeScale, "Time spent in the queue before execution for maintenance actions [ms]");
```

File: arangod/RestServer/QueryRegistryFeature.cpp [hunk 7]
```diff
@@ -97,6 +97,20 @@ namespace arangodb {
 
 std::atomic<aql::QueryRegistry*> QueryRegistryFeature::QUERY_REGISTRY{nullptr};
 
+struct QueryRegistryScale {
```
Comment (mpoeter) [inline arangod/RestServer/QueryRegistryFeature.cpp:100]:
```suggestion
struct QueryTimeScale {
```

File: arangod/RestServer/QueryRegistryFeature.cpp [hunk 8]
```diff
@@ -97,6 +97,20 @@ namespace arangodb {
...
+  static log_scale_t<double> scale() { return {2., 0.0, 50.0, 20}; }
+};
+struct QueryRegistryScale2 {
+  static log_scale_t<double> scale() { return {2., 1.0, 2000.0, 10}; }
+};
+
+DECLARE_COUNTER(arangodb_aql_all_query, "Total number of AQL queries finished");
+DECLARE_HISTOGRAM(arangodb_aql_query_time, QueryRegistryScale, "Execution time histogram for all AQL queries [s]");
```
Comment (mpoeter) [inline arangod/RestServer/QueryRegistryFeature.cpp:108]:
```suggestion
DECLARE_HISTOGRAM(arangodb_aql_query_time, QueryTimeScale, "Execution time histogram for all AQL queries [s]");
```

File: arangod/RestServer/QueryRegistryFeature.cpp [hunk 9]
```diff
@@ -97,6 +97,20 @@ namespace arangodb {
 
 std::atomic<aql::QueryRegistry*> QueryRegistryFeature::QUERY_REGISTRY{nullptr};
 
+struct QueryRegistryScale {
+  static log_scale_t<double> scale() { return {2., 0.0, 50.0, 20}; }
+};
+struct QueryRegistryScale2 {
```
Comment (mpoeter) [inline arangod/RestServer/QueryRegistryFeature.cpp:103]:
```suggestion
struct SlowQueryTimeScale {
```

File: arangod/RestServer/QueryRegistryFeature.cpp [hunk 10]
```diff
@@ -97,6 +97,20 @@ namespace arangodb {
...
+  static log_scale_t<double> scale() { return {2., 1.0, 2000.0, 10}; }
+};
+
+DECLARE_COUNTER(arangodb_aql_all_query, "Total number of AQL queries finished");
+DECLARE_HISTOGRAM(arangodb_aql_query_time, QueryRegistryScale, "Execution time histogram for all AQL queries [s]");
+DECLARE_COUNTER(arangodb_aql_slow_query, "Total number of slow AQL queries finished");
+DECLARE_HISTOGRAM(arangodb_aql_slow_query_time, QueryRegistryScale2, "Execution time histogram for slow AQL queries [s]");
```
Comment (mpoeter) [inline arangod/RestServer/QueryRegistryFeature.cpp:110]:
```suggestion
DECLARE_HISTOGRAM(arangodb_aql_slow_query_time, SlowQueryTimeScale, "Execution time histogram for slow AQL queries [s]");
```

**Review by kvahed — COMMENTED** [review]
LGTM

**Review by kvahed — APPROVED** [review]
LGTM

File: Documentation/Metrics/arangodb_heartbeat_send_time_msec.yaml [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+  Time required to send a heartbeat.
+unit: ms
+type: histogram
+category: Health
+complexity: medium
+exposedBy:
+  - coordinator
+  - dbserver
+description: |
+  Time required to send a heartbeat.
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply Documentation/Metrics/arangodb_heartbeat_send_time_msec.yaml]: Clarified.

[the remark this answers is not in the document]
Comment (mpoeter) [reply Documentation/Metrics/arangodb_heartbeat_send_time_msec.yaml]: My bad, I did not realize that this was a histogram. Anyway, the new description is clearer now! 👍

**Pull request merged**